glib/src/value.rs

```diff
@@ -80,44 +76,216 @@
 //! ```
 
 use libc::{c_char, c_void};
-use std::borrow::Borrow;
 use std::error;
 use std::ffi::CStr;
 use std::fmt;
-use std::marker::PhantomData;
 use std::mem;
 use std::ops::Deref;
 use std::ptr;
 
 use crate::gstring::GString;
 use crate::translate::*;
 use crate::types::{StaticType, Type};
+use crate::{ParamFlags, ParamSpec};
+
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> /* TODO: TryFrom<Value> + Into<Value> */ {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling nullable types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is nullable.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeNullable:
```

Comment:
```suggestion
pub trait ValueTypeOptional:
```

Maybe that's better / more consistent? Also for the other usages of "nullable" in here

Reply:
Yes! Initially, I wrote this exact suggestion and I stepped back when I understood it was intended for expressing glib C null-ability. But since it will be visible within Rust code only, I think `Optinal` is better.